libraries/AP_Compass/AP_Compass.cpp

```diff
@@ -877,6 +881,41 @@ void Compass::_detect_backends(void)
     #error Unrecognised HAL_COMPASS_TYPE setting
 #endif
 
+
+#ifdef BOARD_I2C_BUS_EXT
```

Comment:
I get the reason, but two points:

- the external compasses should probably be detected before the internal compass
- we should this correctly, read #6748 and #6265 (there are probably other comments elsewhere); with that implemented, there would be no reason to not probe all drivers for external buses for all boards, which would greatly reduce the code duplication.

Reply:
> should probably be detected before the internal compass

but why? They are completely independent

>read #6748 and #6265

those PRs are hangs for half of year, while my users lost the ability to update right now, so I want to restore full functionality

Reply:
> but why? They are completely independent

not quite, the EKF uses the first one in preference, and external compass is usually preferred (otherwise why did the user bother to attach it?)

Reply:
>why did the user bother to attach it

because most boards don't contains internal  one. With per-motor compensation builtin compass gives very good results